rclcpp_components/doc/architecture.rst

```diff
@@ -0,0 +1,263 @@
+An example of this is the ``rclcpp_lifecycle::LifecycleNode`` class.
+
+The interface of the ``rclcpp_components::NodeFactory`` class is something like this:
+
+.. code-block:: cpp
+
+    class NodeFactory
+    {
+    public:
+      NodeFactory() = default;
+      virtual ~NodeFactory() = default;
+
+      virtual
+      NodeInstanceWrapper
+      create_node_instance(rclcpp::NodeOptions options) = 0;
+    };
+
+The ``NodeInstanceWrapper`` type would just encapsulate the instance of the node as a type-erased shared pointer, and a method for accessing the ``rclcpp::node_interfaces::NodeBaseInterface`` pointer from that node instance.
+It might look like this:
+
+.. code-block:: cpp
+
+    class NodeInstanceWrapper
+    {
+    public:
+      using NodeBaseInterfaceGetter = std::function<
+        rclcpp::node_interfaces::NodeBaseInterface::SharedPtr (const std::shared_ptr<void> &)
+      >;
+
+      NodeInstanceWrapper(
+        std::shared_ptr<void> node_instance,
+        NodeBaseInterfaceGetter node_base_interface_getter)
+      : node_instance_(node_instance), node_base_interface_getter_(node_base_interface_getter)
+      {}
+
+      const std::shared_ptr<void>
+      get_node_instance() const
+      {
+        return node_instance_;
+      }
+
+      rclcpp::node_interfaces::NodeBaseInterface::SharedPtr
+      get_node_base_interface()
+      {
+        return node_base_interface_getter_(node_instance_);
+      }
+
+    private:
+      std::shared_ptr<void> node_instance_;
+      NodeBaseInterfaceGetter node_base_interface_getter_;
+    };
+
+This interface only requires that you can give access to the ``rclcpp::node_interfaces::NodeBaseInterface`` pointer, as that's the only thing required to compose (and then execute) composable nodes.
+The ability to get access to the original Node instance shared pointer is just for debugging and special cases.
+Most of the time the program loading the instance of the node will not be aware of the original type of the node instance, whether it be a ``rclcpp::Node`` or something like ``my_ns::MyNode`` which inherits from ``rclcpp::Node`` or even something else completely.
+
+Finally, the ``rclcpp_components::NodeFactoryTemplate<NodeT>`` convenience class will create an implementation of ``NodeFactory`` based on the ``NodeT`` type.
+This template class will work so long as the class given for ``NodeT`` provides the ``get_node_base_interface()`` method that exists on the ``rclcpp::Node`` class.
+
+Building the Shared Library which Contains Composable Nodes
+-----------------------------------------------------------
+
+Typically this is done by just calling the standard ``add_library`` macro in CMake, but any thing that results in generating a CMake target which represents a shared library is sufficient.
+
+Once you've created the target you may manipulate the target with standard CMake macros like ``target_include_directories`` and ``target_link_libraries``.
+
+Creating the Plugin Description File
```

Comment:
Just taking https://github.com/ros2/demos/tree/rolling/composition as an example, I don't see any pluginlib XML description file, either in the sources nor in the `install` directory.  But maybe I'm missing something.